src/goto-instrument/synthesizer/expr_enumerator.h

```diff
@@ -0,0 +1,511 @@
+      for(const auto &product_tuple :
+          cartesian_product_of_enumerators(sub_enumerators))
+      {
+        // Optimization: rule out equivalent expressions
+        // using certain equivalence class.
+        // Keep only representation tuple of each equivalence class.
+        if(is_equivalence_class_representation(product_tuple))
+          result.push_back(instantiate(product_tuple));
+      }
+    }
+
+    return result;
+  }
+
+  void set_size(unsigned int new_size) override
+  {
+    expected_size = new_size;
+  }
+
+  non_leaf_enumeratort *clone() const override
+  {
+    return new non_leaf_enumeratort(sub_enumerators, is_none_enumerator);
+  }
+
+  /// As default, keep all expression tuples.
+  virtual bool is_equivalence_class_representation(const exprst &es)
+  {
+    return true;
+  }
+
+protected:
+  /// Propogate partition to sub-enumerators.
+  void set_sub_enumerator_sizes(const partitiont &sizes)
+  {
+    INVARIANT(sizes.size() == arity, "arity not equal to number of arguments");
+    auto front = sizes.begin();
+    for(auto &sub_enumerator : sub_enumerators)
+    {
+      sub_enumerator->set_size(*front);
+      std::advance(front, 1);
+    }
+  }
+
+  /// Instantiate a list of expressions to a expression using
+  /// a top-level operator.
+  virtual exprt instantiate(const exprst &exprs)
+  {
+    return true_exprt();
+  }
+
+  unsigned int arity;
+  unsigned int expected_size;
+  enumeratorst sub_enumerators;
+  std::function<bool(const partitiont &)> is_good_partition;
+};
+
+/// Enumerator that enumerates binary functional expressions.
+class binary_functional_enumeratort : public non_leaf_enumeratort
+{
+public:
+  binary_functional_enumeratort(
+    const irep_idt &op,
+    const enumerator_baset &enumerator_1,
+    const enumerator_baset &enumerator_2,
+    std::function<bool(const partitiont &)> partition_check)
+    : op_id(op)
+  {
+    is_none_enumerator = false;
+    op_id = op;
+    arity = 2;
+    sub_enumerators = enumeratorst();
+    sub_enumerators.push_back(enumerator_1.clone());
+    sub_enumerators.push_back(enumerator_2.clone());
+    is_good_partition = partition_check;
+  }
+
+  binary_functional_enumeratort(
+    const irep_idt &op,
+    const enumerator_baset &enumerator_1,
+    const enumerator_baset &enumerator_2)
+    : binary_functional_enumeratort(
+        op,
+        enumerator_1,
+        enumerator_2,
+        [](const partitiont &) { return true; })
+  {
+  }
+
+  void set_size(unsigned int new_size) override
+  {
+    expected_size = new_size;
+  }
+
+  void set_non_exchangable()
+  {
+    is_exchangable = false;
+  }
+
+  binary_functional_enumeratort *clone() const override
+  {
+    return new binary_functional_enumeratort(
+      op_id,
+      *(sub_enumerators.front()->clone()),
+      *(sub_enumerators.back()->clone()),
+      is_good_partition);
+  }
+
+  bool is_commutative(const irep_idt op)
+  {
+    return op_id == ID_equal || op_id == ID_plus;
+  }
```

Comment:
What about multiplication, not-equal, and, or, bitand, bitor, bitxor? You might want to look at, or perhaps even call the functions from src/util/simplify_utils.cpp.